### PR TITLE
Multi Oracle Support!

### PR DESCRIPTION
**Pull request by nkohen**
Replaces #2314

Built on top of #2505

Implements arbitrary multi-oracle support for DLCs!

Needs tests & documentation, as well as some simple refactoring and optimization.

### CODE REVIEW

**Comment by nkohen** [comment]
Ah shoot, note to self: `Vector::prepended` has a different name in scala 2.12 so you should just use `Vector(_) ++ _`

File: build.sbt [hunk 1]
```diff
@@ -106,6 +106,12 @@ lazy val `bitcoin-s` = project
     feeProviderTest,
     dlcOracle,
     dlcOracleTest,
+    dlc,
+    dlcTest,
+    dlcWallet,
+    dlcWalletTest,
+    dlcSuredbitsClient,
+    dlcSuredbitsClientTest,
```
Comment (nkohen) [inline build.sbt:114]: what's this needed for?

File: dlc/src/main/scala/org/bitcoins/dlc/testgen/DLCParsingTestVector.scala [hunk 2]
```diff
@@ -449,6 +451,14 @@ object DLCParsingTestVector extends TestVectorParser[DLCParsingTestVector] {
         )
 
         DLCMessageTestVector(LnMessage(tlv), "oracle_announcement_v0", fields)
+      case OracleAttestmentV0TLV(pubkey, sigs) =>
+        val fields = Vector(
+          "tpe" -> Element(UInt16(OracleAttestmentV0TLV.tpe.toInt)),
```
Comment (nkohen) [inline dlc/src/main/scala/org/bitcoins/dlc/testgen/DLCParsingTestVector.scala:457]: needs length

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/CETCalculator.scala [hunk 3]
```diff
@@ -402,4 +405,370 @@ object CETCalculator {
...
+  def computeCETIntervalBinary(
+      cet: Vector[Int],
+      numDigits: Int): (Long, Long) = {
+    val left = NumberUtil.fromDigits(cet, base = 2, numDigits)
+    val right = left + (1L << (numDigits - cet.length)) - 1
+
+    (left, right)
+  }
+
+  // TODO: Delete this when merging
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/CETCalculator.scala]: ping

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/CETCalculator.scala [hunk 4]
```diff
@@ -402,4 +405,370 @@ object CETCalculator {
...
+      numDigits: Int): Vector[Int] = {
+    val leftBound = start - minFail
+    val leftBoundDigits = NumberUtil.decompose(leftBound, base = 2, numDigits)
+    val (prefix, halvingDigits) =
+      leftBoundDigits.splitAt(numDigits - maxErrorExp)
+
+    prefix ++ halvingDigits.takeWhile(_ == 1)
+  }
+
+  private def singleCoveringCETCombinations[T](
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/CETCalculator.scala:503]: ping (and below)

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractDescriptor.scala [hunk 5]
```diff
@@ -0,0 +1,120 @@
+package org.bitcoins.core.protocol.dlc
+
+import org.bitcoins.core.currency.Satoshis
+import org.bitcoins.core.protocol.tlv._
+import org.bitcoins.core.util.SeqWrapper
+
+sealed trait ContractDescriptor extends TLVSerializable[ContractDescriptorTLV] {
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractDescriptor.scala:16]: scaladoc

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractDescriptor.scala [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+  override def fromTLV(tlv: ContractDescriptorTLV): ContractDescriptor = {
+    tlv match {
+      case tlv: ContractDescriptorV0TLV => EnumContractDescriptor.fromTLV(tlv)
+      case tlv: ContractDescriptorV1TLV =>
+        NumericContractDescriptor.fromTLV(tlv)
+    }
+  }
+}
+
+case class EnumContractDescriptor(
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractDescriptor.scala:48]: scaladoc

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+  ContractInfoV0TLV,
+  TLVDeserializable,
+  TLVSerializable,
+  UnsignedNumericOutcome
+}
+import org.bitcoins.crypto.ECPublicKey
+
+import scala.collection.immutable.HashMap
+
+case class ContractInfo(
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala:25]: scaladoc

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+  }
+
+  /** Returns the maximum payout this party could win from this contract */
+  val max: Satoshis = contractDescriptor match {
+    case descriptor: EnumContractDescriptor =>
+      descriptor.values.maxBy(_.toLong)
+    case _: NumericContractDescriptor => totalCollateral
+  }
+
+  val descriptorAndInfo: Either[
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala:49]: scaladoc

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+        Left((contractDescriptor, oracleInfo))
+      case (contractDescriptor: NumericContractDescriptor,
+            oracleInfo: NumericOracleInfo) =>
+        Right((contractDescriptor, oracleInfo))
+      case (_: ContractDescriptor, _: OracleInfo) =>
+        throw new IllegalArgumentException(
+          s"All infos must be for the same kind of outcome: $this")
+    }
+
+  /** Vector is always the most significant digits */
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala]: scaladoc

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+        val adaptorPoint = outcome.sigPoint
+
+        builder.+=(outcome -> (adaptorPoint, offerPayout, acceptPayout))
+    }
+
+    builder.result()
+  }
+
+  // TODO: Needs a lot of optimization
+  def verifySigs(
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala:176]: scaladoc

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+            oraclesAndOutcomes.foldLeft(true) {
+              case (boolSoFar, (oracle, digits)) =>
+                lazy val sig = sigs.find(_.oracle == oracle)
+                boolSoFar && sig.exists(_.verifySignatures(digits))
+            }
+        }
+    }
+  }
+
+  def findOutcome(sigs: Vector[OracleSignatures]): Option[OracleOutcome] = {
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala:211]: scaladoc

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+    // TODO: Optimize by looking at nonces
+    // TODO: Optimize using NumericOracleSignatures.computeOutcome
+    allOutcomes.find(verifySigs(_, sigs))
+  }
+
+  def resultOfOutcome(
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala:218]: scaladoc

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+  }
+
+  /** Returns the payouts for the outcome as (toOffer, toAccept) */
+  def getPayouts(outcome: OracleOutcome): (Satoshis, Satoshis) = {
+    val (_, offerOutcome, acceptOutcome) = resultOfOutcome(outcome)
+
+    (offerOutcome, acceptOutcome)
+  }
+
+  def updateOnAccept(
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/ContractInfo.scala:248]: scaladoc

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/OracleInfo.scala [hunk 14]
```diff
@@ -0,0 +1,291 @@
+package org.bitcoins.core.protocol.dlc
+
+import org.bitcoins.core.protocol.tlv._
+import org.bitcoins.crypto._
+
+sealed trait OracleInfo extends TLVSerializable[OracleInfoTLV] {
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/OracleInfo.scala:9]: scaladocs

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/OracleOutcome.scala [hunk 15]
```diff
@@ -0,0 +1,78 @@
+package org.bitcoins.core.protocol.dlc
+
+import org.bitcoins.core.protocol.tlv.{
+  DLCOutcomeType,
+  EnumOutcome,
+  UnsignedNumericOutcome
+}
+import org.bitcoins.crypto.{ECPublicKey, SchnorrNonce}
+
+sealed trait OracleOutcome {
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/OracleOutcome.scala:15]: scaladocs

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/OracleSignatures.scala [hunk 16]
```diff
@@ -0,0 +1,96 @@
+package org.bitcoins.core.protocol.dlc
+
+import org.bitcoins.core.protocol.tlv.{DLCOutcomeType, UnsignedNumericOutcome}
+import org.bitcoins.core.util.SeqWrapper
+import org.bitcoins.crypto.{CryptoUtil, ECPrivateKey, SchnorrDigitalSignature}
+
+sealed trait OracleSignatures extends SeqWrapper[SchnorrDigitalSignature] {
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/OracleSignatures.scala:8]: scaladocs

File: dlc/src/main/scala/org/bitcoins/dlc/execution/DLCExecutor.scala [hunk 17]
```diff
@@ -68,24 +73,30 @@ case class DLCExecutor(signer: DLCTxSigner)(implicit ec: ExecutionContext) {
   }
 
   /** Return's this party's payout for a given oracle signature */
-  def getPayout(sigs: Vector[SchnorrDigitalSignature]): CurrencyUnit = {
+  def getPayout(sigs: Vector[OracleSignatures]): CurrencyUnit = {
     signer.getPayout(sigs)
   }
 
   def executeDLC(
```
Comment (nkohen) [inline dlc/src/main/scala/org/bitcoins/dlc/execution/DLCExecutor.scala:85]: scaladoc

File: testkit/src/main/scala/org/bitcoins/testkit/dlc/DLCTest.scala [hunk 18]
```diff
@@ -0,0 +1,793 @@
...
+      acceptSetup.cets.foreach {
+        case (msg, cetInfo) =>
+          assert(cetInfo.tx == offerSetup.cets(msg).tx)
+      }
+
+      (offerSetup, acceptSetup)
+    }
+  }
+
+  def genEnumOracleSignature(
```
Comment (nkohen) [inline testkit/src/main/scala/org/bitcoins/testkit/dlc/DLCTest.scala:503]: scaladocs

**Comment by nkohen** [comment]
@benthecarman looks like dlc wallet tests are timing out

File: build.sbt (hunk 1 above)
Comment (benthecarman) [reply build.sbt:114]: @Christewart added all the modules in the `dependsOn` for intellij support I think. This adds the DLC branch exclusive ones to it

File: testkit/src/main/scala/org/bitcoins/testkit/dlc/DLCTest.scala [hunk 19]
```diff
@@ -0,0 +1,773 @@
...
+    val sig = privKey.schnorrSignWithNonce(CryptoUtil
+                                             .sha256DLCAttestation(outcome)
+                                             .bytes,
+                                           kVal)
+
+    EnumOracleSignature(oracleInfo, sig)
+  }
+
+  def genEnumOracleOutcome(
+      chosenOracles: Vector[Int],
```
Comment (benthecarman) [inline testkit/src/main/scala/org/bitcoins/testkit/dlc/DLCTest.scala:517]: With over-sharing of oracles added, should make it sometimes use a larger set of oracles

**Review by benthecarman — APPROVED** [review]
Overall lgtm barring all the optimization todos are done in a followup

File: testkit/src/main/scala/org/bitcoins/testkit/dlc/DLCTest.scala (hunk 19 above)
Comment (nkohen) [reply testkit/src/main/scala/org/bitcoins/testkit/dlc/DLCTest.scala:517]: added a todo

**Pull request merged**